src/git-dep.stanza

```diff
@@ -77,64 +78,42 @@ public defn parse-git-dependency (name:String, locator:String, version?:Maybe<St
   val hash = "" ; This gets resolved during the fetch/sync
   GitDependency(name, locator, version, hash)
 
-public defn fetch-or-sync-at-hash (d: GitDependency) -> False:
-  if not has-git?() :
-    error("No `git` executable found on the path")
-
-  if file-exists?(path(d)):
-    sync-dependency-to-hash(d)
-  else:
-    fetch-dependency-at-hash(d)
-
-
-public defn sync-dependency-to-hash (d: GitDependency) -> False:
-  debug("syncing '%_' at %_" % [name(d), hash(d)])
-
-  if empty?(hash(d)): fatal("internal error")
-
-  if git-rev-parse!(path(d), "HEAD") != hash(d):
-    error("syncing '%_': out of sync with your slm.lock" % [name(d)])
-
-public defn fetch-dependency-at-hash (d: GitDependency) -> False:
-  info("cloning %_ at %_" % [colored-name?(d), hash(d)])
-
-  if empty?(hash(d)): fatal("internal error")
-
-  val url = full-url-from-locator(locator(d))
-  shallow-clone-git-repo(url, path(d))
-
-  run-git-command-in-dir(path(d), ["fetch", "--quiet", "origin", hash(d)])
-  run-git-command-in-dir(path(d), ["checkout", "--quiet", "--force", hash(d)])
-
-  false
-
-
 public defn fetch-dependency-at-version (d: GitDependency) -> False:
   info("cloning %_ at %~" % [colored-name?(d), colored-version?(d)])
-
-  val url = full-url-from-locator(locator(d))
-  shallow-clone-git-repo(url, path(d))
-
-  val tag = to-string("v%~" % [version(d)])
-  val refspec = to-string("+refs/tags/%_:refs/tags/%_" % [tag, tag])
-  run-git-command-in-dir(path(d), ["fetch", "--quiet", "origin", refspec])
-  run-git-command-in-dir(path(d), ["checkout", "--quiet", "--force", tag])
-
+  ; Use git if available
+  if has-git?() :
```

Comment:
You do this pattern a lot, ie - If we have the `git` binary - use it. Other wise - use `libgit2`. 

I posit that:

1.  This puts a lot of the nuts and bolts of the git clone/fetch/sync/whatever in the application code of `slm`
1.  The interface to `git` is quite well defined for our purposes.
2.  We can structure this slightly differently to make the abstract easier to use. 

For example - we could make a `GitInterface` base type with the following types of methods: 

```
deftype GitInterface
defmulti fetch(tag:String) -> False 
defmulti sync(tag:String) -> False 
defmulti rev-parse(...) -> False
...
```

Then we can implement two concrete types for the `git` binary version and the `libgit2` version. 

Thoughts ? Any reason not to do this?

Reply:
That is what I initially tried to do, and would certainly look nicer. However, doing so would require repeatedly re-initializing `libgit` and re-opening the repository for each individual operation. I could try to create a "LibGitSession" object of sorts to avoid that, but passing that object around would require additional plumbing which the `git` binary version does not have. If we don't mind the performance overhead of re-initializing and re-opening for each operation I can change it.

Reply:
Can you please provide a more concrete example of where this is failing?

From my perspective - implementing as an interface like this allows us to hide all of those details and make the application layer code for SLM more maintainable.